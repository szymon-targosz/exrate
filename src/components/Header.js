import React from 'react'

const Header = () => (
   <header className="header">
      <div className="container">
         <h1 className="header__title">Exrate</h1>
      </div>
   </header>
)

export default Header;
