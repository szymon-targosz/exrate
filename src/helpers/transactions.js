import numeral from 'numeral';
import 'numeral/locales';

const getHighest = transactions =>
   transactions.length === 0 ? undefined : transactions.reduce((prev, curr) => (prev.amount > curr.amount ? prev : curr), []);

const getTotal = transactions => transactions.map(transaction => transaction.amount).reduce((sum, val) => sum + val, 0);

const selectTransactions = (transactions, { text, sort }) => {
   return transactions
      .filter(transaction => {
         const textMatch = transaction.name.toLowerCase().includes(text.toLowerCase());
         return textMatch;
      })
      .sort((a, b) => {
         if (sort === 'alpha') {
            return a.name > b.name ? 1 : -1;
         } else if (sort === 'asc') {
            return a.amount - b.amount;
         } else {
            return b.amount - a.amount;
         }
      });
};

const generateAmountMsg = (amount, locale = 'de', decsFit = false) => {
   numeral.locale(locale);

   let decNums = '00';
   if (decsFit) {
      const decs = amount.toString().split('.')[1] || '';
      let i = 0;
      while (i < decs.length - 2) {
         decNums += 0;
         i++;
      }
   }

   return numeral(amount).format(`0,0.${decNums} $`);
};

export { getHighest, getTotal, selectTransactions, generateAmountMsg };
