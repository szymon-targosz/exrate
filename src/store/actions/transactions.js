import uuidv4 from 'uuid/v4';
import { ADD_TRANSACTION, REMOVE_TRANSACTION, REMOVE_ALL_TRANSACTIONS } from './actionTypes';

const addTransaction = (name, amount) => ({
   type: ADD_TRANSACTION,
   transaction: {
      id: uuidv4(),
      amount: parseFloat(amount),
      name
   }
});

const removeTransaction = id => ({
   type: REMOVE_TRANSACTION,
   id
});

const removeAllTransactions = () => ({ type: REMOVE_ALL_TRANSACTIONS });

export { addTransaction, removeTransaction, removeAllTransactions };
