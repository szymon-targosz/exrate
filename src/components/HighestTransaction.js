import React from 'react';
import { connect } from 'react-redux';
import { getHighest, selectTransactions } from '../helpers/transactions';
import Transaction from './Transaction';
import SectionHeader from './UI/SectionHeader';

export const HighestTransaction = ({ transaction, rate }) => {
   const jsx = (
      <section className="highest">
         <SectionHeader title="Highest transaction" />
         <Transaction transaction={transaction} rate={rate} />
      </section>
   );

   return transaction ? jsx : null;
};

const mapStateToProps = ({ transactions, rate, filters }) => ({
   transaction: getHighest(selectTransactions(transactions, filters)),
   rate
});

export default connect(mapStateToProps)(HighestTransaction);
