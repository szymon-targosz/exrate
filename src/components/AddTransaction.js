import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addTransaction } from '../store/actions';
import SectionHeader from './UI/SectionHeader';
import Input from './UI/Input';

export class AddTransaction extends Component {
   state = { name: '', amount: '', error: undefined };
   inputNameRef = React.createRef();

   handleSubmit = e => {
      e.preventDefault();
      const name = this.state.name.trim();
      if (name === '') return this.setState({ error: 'Please enter the transaction name' });

      const isOccupied = this.props.transactions.find(t => t.name === name);
      if (isOccupied) return this.setState({ error: 'Name already taken' });

      this.props.addTransaction(name, this.state.amount);
      this.setState({ error: undefined, name: '', amount: '' });
      this.inputNameRef.current.focus();
   };

   onNameChange = e => {
      this.setState({ name: e.target.value });
   };

   onAmountChange = e => {
      const amount = e.target.value;
      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
         this.setState({ amount });
      }
   };

   render() {
      return (
         <section className="add-transaction">
            <SectionHeader title="Add transaction" />
            <form onSubmit={this.handleSubmit} className="p-s">
               <Input
                  elementType="input"
                  changeHandler={this.onNameChange}
                  config={{
                     name: 'name',
                     ref: this.inputNameRef,
                     value: this.state.name,
                     placeholder: 'Name',
                     autoComplete: 'off',
                     id: 'name-input',
                     type: 'text'
                  }}
                  label
               />
               <Input
                  elementType="input"
                  changeHandler={this.onAmountChange}
                  config={{
                     name: 'amount',
                     value: this.state.amount,
                     placeholder: 'Amount (EUR)',
                     autoComplete: 'off',
                     id: 'amount-input',
                     type: 'text'
                  }}
                  label
               />
               {this.state.error && <p className="error">{this.state.error}</p>}
               <button className="btn btn--block" disabled={!this.state.name || !this.state.amount}>
                  Add
               </button>
            </form>
         </section>
      );
   }
}

const mapStateToProps = ({ transactions }) => ({ transactions });

const mapDispatchToProps = dispatch => ({
   addTransaction: (name, amount) => dispatch(addTransaction(name, amount))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTransaction);
