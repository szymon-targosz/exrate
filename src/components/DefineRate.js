import React, { Component } from 'react';
import { connect } from 'react-redux';
import SectionHeader from './UI/SectionHeader';
import InfoBox from './UI/InfoBox';
import Input from './UI/Input';
import { setExchangeRate } from '../store/actions';
import { generateAmountMsg } from '../helpers/transactions';

export class DefineRate extends Component {
   state = { rate: 0 };

   handleInput = e => {
      const rate = e.target.value;
      if (!rate || rate.match(/^\d{1,}(\.\d{0,9})?$/)) {
         this.setState({ rate: rate });
         this.props.setExchangeRate(rate || 0);
      }
   };

   render() {
      return (
         <section className="define-rate">
            <SectionHeader title="Set exchange rate" cssClasses="define-rate__header" />
            <div className="define-rate__form p-s">
               <Input
                  elementType="input"
                  cssClasses="input__control--short"
                  changeHandler={this.handleInput}
                  config={{
                     value: this.state.rate,
                     placeholder: 'Exchange rate',
                     id: 'rate-input',
                     type: 'text'
                  }}
                  label
               />
            </div>
            <div className="define-rate__current">
               <InfoBox
                  description="Current exchange rate"
                  data={`${generateAmountMsg(1)} = ${generateAmountMsg(parseFloat(this.state.rate) || 0, 'pl', true)}`}
               />
            </div>
         </section>
      );
   }
}

const mapDispatchToProps = dispatch => ({ setExchangeRate: rate => dispatch(setExchangeRate(rate)) });

export default connect(null, mapDispatchToProps)(DefineRate);
