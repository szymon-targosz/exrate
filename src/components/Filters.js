import React, { Component } from 'react';
import SectionHeader from './UI/SectionHeader';
import Input from './UI/Input';
import { connect } from 'react-redux';
import { setTextFilter, sortAlpha, sortAsc, sortDesc } from '../store/actions';

export class Filters extends Component {
   onTextChange = e => {
      this.props.setTextFilter(e.target.value);
   };

   onSortChange = e => {
      const sort = e.target.value;
      if (sort === 'alpha') {
         this.props.sortAlpha();
      } else if (sort === 'desc') {
         this.props.sortDesc();
      } else {
         this.props.sortAsc();
      }
   };

   render() {
      return (
         <section className="filters">
            <SectionHeader title="Filters" />
            <div className="p-s">
               <Input
                  elementType="input"
                  config={{
                     placeholder: 'Search',
                     value: this.props.filters.text,
                     id: 'search-input',
                     type: 'search',
                     autoComplete: 'off'
                  }}
                  changeHandler={this.onTextChange}
                  label
               />
               <Input
                  elementType="select"
                  cssClasses="input__control--select"
                  config={{ value: this.props.filters.sort }}
                  changeHandler={this.onSortChange}
                  options={[
                     { value: 'alpha', text: 'Alphabetically' },
                     { value: 'asc', text: 'Ascending' },
                     { value: 'desc', text: 'Descending' }
                  ]}
               />
            </div>
         </section>
      );
   }
}

const mapStateToProps = ({ filters }) => ({ filters });

const mapDispatchToProps = dispatch => ({
   setTextFilter: text => dispatch(setTextFilter(text)),
   sortAlpha: () => dispatch(sortAlpha()),
   sortAsc: () => dispatch(sortAsc()),
   sortDesc: () => dispatch(sortDesc())
});

export default connect(mapStateToProps, mapDispatchToProps)(Filters);
