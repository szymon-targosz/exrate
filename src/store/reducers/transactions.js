import { ADD_TRANSACTION, REMOVE_TRANSACTION, REMOVE_ALL_TRANSACTIONS } from '../actions/actionTypes';

export default (state = [], action) => {
   switch (action.type) {
      case ADD_TRANSACTION:
         return state.concat(action.transaction);

      case REMOVE_TRANSACTION:
         return state.filter(({ id }) => id !== action.id);

      case REMOVE_ALL_TRANSACTIONS:
         return [];

      default:
         return state;
   }
};
