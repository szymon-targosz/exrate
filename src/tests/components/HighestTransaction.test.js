import React from 'react';
import { HighestTransaction } from '../../components/HighestTransaction';
import { shallow } from 'enzyme';
import transactions from '../fixtures/transactions';

describe('COMPONENT - HighestTransaction', () => {
   test('Should render HighestTransaction correctly', () => {
      const wrapper = shallow(<HighestTransaction transaction={transactions[0]} rate={4.46} />);
      expect(wrapper).toMatchSnapshot();
   });
   
   test('Should not render HighestTransaction if there is no transaction', () => {
      const wrapper = shallow(<HighestTransaction transaction={undefined} rate={4.46} />);
      expect(wrapper).toMatchSnapshot();
   });
});
