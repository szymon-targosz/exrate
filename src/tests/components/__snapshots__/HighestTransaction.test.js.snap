// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - HighestTransaction Should not render HighestTransaction if there is no transaction 1`] = `""`;

exports[`COMPONENT - HighestTransaction Should render HighestTransaction correctly 1`] = `
<section
  className="highest"
>
  <SectionHeader
    title="Highest transaction"
  />
  <Transaction
    rate={4.46}
    transaction={
      Object {
        "amount": 41,
        "id": "1",
        "name": "Bills",
      }
    }
  />
</section>
`;
